controllers/machinedeletionremediation_controller.go

```diff
@@ -76,73 +99,125 @@ type MachineDeletionRemediationReconciler struct {
 //
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
-func (r *MachineDeletionRemediationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+func (r *MachineDeletionRemediationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (finalResult ctrl.Result, finalErr error) {
 	log := r.Log.WithValues("machinedeletionremediation", req.NamespacedName)
 
 	log.Info("reconciling...")
 
 	var err error
 	var remediation *v1alpha1.MachineDeletionRemediation
 	if remediation, err = r.getRemediation(ctx, req); remediation == nil || err != nil {
-		if apiErrors.IsNotFound(err) {
-			return ctrl.Result{}, nil
-		}
 		return ctrl.Result{}, err
 	}
 
-	annotations := remediation.GetAnnotations()
-	if annotations != nil {
-		machineNameNamespace, exists := annotations[MachineNameNamespaceAnnotation]
-		if exists {
-			machineName, machineNamespace, err := extractNameAndNamespace(machineNameNamespace, remediation.GetName())
-			if err != nil {
-				log.Error(err, "could not get Machine data from remediation", "remediation", remediation.GetName(), "annotation", machineNameNamespace)
-				return ctrl.Result{}, nil
-			}
-
-			ok, err := r.verifyMachineIsDeleted(ctx, machineName, machineNamespace, remediation.Name)
-			if err != nil {
-				return ctrl.Result{}, err
-			}
-			if !ok {
-				return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
+	defer func() {
+		if updateErr := r.updateStatus(ctx, remediation); updateErr != nil {
+			if !apiErrors.IsConflict(updateErr) {
+				finalErr = utilerrors.NewAggregate([]error{updateErr, finalErr})
+			} else {
+				// requeue if there is no error already set
+				if finalErr == nil {
+					// if requeue is already set choose the lowest delay
+					minRequeue := time.Second
+					if finalResult.RequeueAfter > 0 && minRequeue > finalResult.RequeueAfter {
+						minRequeue = finalResult.RequeueAfter
+					}
+					finalResult.RequeueAfter = minRequeue
+				}
 			}
-			return ctrl.Result{}, nil
 		}
-	}
+	}()
 
-	var node *v1.Node
-	if node, err = r.getNodeFromMdr(remediation); err != nil {
-		return ctrl.Result{}, err
+	if r.isStoppedByNHC(remediation) {
+		log.Info("MDR remediation was stopped by Node Healthcheck")
+		return ctrl.Result{}, r.updateConditions(remediationTerminatedByNHC, remediation)
 	}
 
-	log.Info("MDR CR and affected Node found", "node", node.GetName())
+	state := r.getCurrentState(remediation)
+	switch state {
+	case mdrStateNotStarted:
+		// check machine has controller owner
+		// set status.condition remediation started
+		node, machine, err := r.getNodeAndMachineFromCR(remediation)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		log.Info("MDR CR and affected Node found", "node", node.GetName())
```

Comment:
NIT: MDR CR and affected Node found -> MDR CR and affected Node was found

Reply:
I have missed that, but I think there is no need to change it